### PR TITLE
Dev IO - Socket Custom Props Import/Export (WIP)

### PR DESCRIPTION
**Pull request by DolphinDream**
As per @zeffii suggestion (PR #1561) the IO tools were extended to export/import socket custom properties.

Note: This branch is branched off the socketExpander branch developed/discussed in PR #1561

With this addition the socket prop (used via use_prop) as well as newly added custom props (use_expander, use_quicklink and expanded) are exported and imported properly via gist.

More testing is necessary to prove it works in all case + better/more robust implementation + cleanup is needed. E.g. the exporting/importing of the socket “prop may have to be done via a proper serialize/deserialize prop method.

### CODE REVIEW

**Comment by DolphinDream** [comment]
Btw, the import/export via gist saves the nodes custom color (the one user can set via property panel) however the bool “use custom color” property doesn’t seem to be exported/imported. I wonder if this should be added as a custom property to be exported/imported via gist ?

File: utils/sv_IO_panel_tools.py [hunk 1]
```diff
@@ -272,11 +272,44 @@ def create_dict_of_tree(ng, skip_set={}, selected=False):
...
+                    value = getattr(socket, tracked_prop)
+
+                    print("Processing custom property: ", tracked_prop, " value = ", value)
+
+                    storable[tracked_prop] = value
+
+                    if tracked_prop == 'use_prop' and value:
+                        print("prop type:", type(socket.prop))
+                        storable['prop'] = socket.prop[:]
+                        # storable['socket_prop_value'] = socket.prop[:]
```
Comment (zeffii) [inline utils/sv_IO_panel_tools.py]: keep it explicit, verbose even.
``` python
storable['socket_prop_value'] = socket.prop[:] 
```

**Comment by zeffii** [comment]
once this is pushed , i will convert all printstatements in `io_panel_tools`  to use the `logging` module, which will check if IO_DEBUG_LEVEL is 0 == no printing, 1 == some printing 2 == kill me already with the printing.

**Comment by zeffii** [comment]
`.use custom color`  could be stored/restored too, i agree. in superficial props. and only if it's set to True.
```python
def get_superficial_props(node_dict, node):
    node_dict['height'] = node.height
    node_dict['width'] = node.width
    node_dict['label'] = node.label
    node_dict['hide'] = node.hide
    node_dict['location'] = node.location[:]
    node_dict['color'] = node.color[:]
    if node.use_custom_color:
        node_dict['use_custom_color'] = True

```

**Comment by zeffii** [comment]
my suggestion was to branch master directly :

- (i've done it for you) https://github.com/nortikin/sverchok/tree/development_io
- then merge your socket_expander into that  *1
- then merge io fixes into that  *2
- then merge the result back into master

*1  and *2 are easier to test for me if the branch is directly on nortikin/sverchok..

File: utils/sv_IO_panel_tools.py (hunk 1 above)
Comment (DolphinDream) [reply utils/sv_IO_panel_tools.py]: Sure.. but then when you restore from socket_prop_value you have to put that into a prop value. The current implementation actually works. I did use your socket_prop_value initially but the prop value was not set properly. If you prefer an indirection/mapping that’s fine, though I don’t see why we would do that. You are exporting a custom property with the same name as the one stored in the class just like the other props (use_expander etc).

**Comment by DolphinDream** [comment]
Sorry for the excessive printing. I was going to take it out once I figure out what’s going on.

**Comment by DolphinDream** [comment]
Sure, if it’s easier for you to test I’ll follow your branching scheme, otherwise the result code is exactly the same.

**Comment by DolphinDream** [comment]
I’ll add the use_custom_color to superficial props. Btw, why would you only want to export this flag if it is set, but the user defined colors are exported regardless ?

**Comment by zeffii** [comment]
if that bothers you, i understand
```python
    if node.use_custom_color:
        node_dict['color'] = node.color[:]
        node_dict['use_custom_color'] = True
```

**Comment by DolphinDream** [comment]
@zeffii All set. **socketExpander** and **devIO** branched off master and **socketExpander** merged into **devIO**

Btw, I added the use_custom_color code you suggested but the flag doesn’t import properly still. i’ll look more into it later unless you can figure out something in the meantime.

**Comment by zeffii** [comment]
i don't see any pullrequests directed at `nortikin/development_io`

anyway, this function also needs to be modified..
```python
def apply_superficial_props(node, node_ref):
    '''
    copies the stored values from the json onto the new node's corresponding values.
    '''
    props = ['location', 'height', 'width', 'label', 'hide', 'color']
    for p in props:
        setattr(node, p, node_ref[p])
```
maybe
```python
def apply_superficial_props(node, node_ref):
    '''
    copies the stored values from the json onto the new node's corresponding values.
    '''
    props = ['location', 'height', 'width', 'label', 'hide']
    for p in props:
        setattr(node, p, node_ref[p])

    if node_ref.get('use_custom_color'):
        node.use_custom_color = True
        node.color = node_ref['color']
```

**Comment by zeffii** [comment]
am pushing this to `nortikin:development_io` - now more people can test it ...and i'll get to try it for the first time :)

**Comment by zeffii** [comment]
maybe for group sanity, i'd like to propose the following process in future.
- avoid major commits
- but.. if you must...forget about `master` for a while.
    - branch `nortikin:master` into `nortikin:development_feature_name`
    - make your edits in `nortikin:development_feature_name`
    - once everyone with say/input has had fair oppurtunity to test/bug report, 
    - then push from `nortikin:development_feature_name` into `nortikin:master`, use the `merge/squeeze`

**Comment by DolphinDream** [comment]
about the “process” a few clarifications if I may:

1. by “major commits” you mean lots of changes into one commit? or lots of changes in many commits into one PR? or changes that touch fundamentals of SV code that are potentially going to make SV unstable ?
2. If i make changes to fundamentals of SV (e.g not a new node etc) can I make this into a personal development_feature branch  or should I make it directly to nortikin:development_feature branch ? Or perhaps make it to a personal dev branch then merge it into a nortikin dev branch for other people to test it before it’s merged into master ?

In retrospect.. should I have not started a PR for a personal socketExpander branch ? (as this touches fundamental SV code?). As I am also member of nortikin team I dont’ know exactly what are the do’s and dont’s, recommended process etc, so as far as I’m concerned I thought branching off in my SV fork and PR against a nortikin repo was perfectly reasonable, but maybe it’s not. I would think that it’s fairly easy to pull from other forks into a testing branch to test other people’s code before you decide to approve the PR, though I’m not doing a whole lot of that so I can’t say that’s always the case.

**Comment by DolphinDream** [comment]
@zeffii  

> if that bothers you, i understand (about saving the color only when the flag is set)

Actually it doesn't bother me that much.. I simply noticed the inconsistency.

**Comment by DolphinDream** [comment]
btw, I updated first commit message (rebased) which somehow accidentally included a bunch of comments related to the matrix math and spiral node (probably a mix up during merging). No code changes, just one commit message changed. Please git pull this branch before adding more commits and/or testing.

**Comment by DolphinDream** [comment]
@zeffii In reference the **serialize_prop** related comments in #1561 .. do you have a serialize_prop code somewhere in your pocket or was that just psedo code ..  meaning it was left to create a prop serializer/deserializer of sort ?  I understand for now that prop[:] works since this is mainly used for vector props.

**Pull request merged**